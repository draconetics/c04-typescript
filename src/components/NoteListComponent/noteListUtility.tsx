// a little function to help us with reordering the result
export const reorder = (list:INote[], startIndex:number, endIndex:number) => {
    const result = Array.from(list);
    const [removed] = result.splice(startIndex, 1);
    result.splice(endIndex, 0, removed);

    return result;
};

/**
 * Moves an item from one list to another list.
 */
export const move = (source:INote[], destination:INote[], droppableSource:any, droppableDestination:any) => {
    const sourceClone = Array.from(source);
    const destClone = Array.from(destination);
    const [removed] = sourceClone.splice(droppableSource.index, 1);

    destClone.splice(droppableDestination.index, 0, removed);

    if(droppableSource.droppableId === "droppable"){
        return {"droppable":sourceClone,"droppable2":destClone};    
    }else{
        return {"droppable":destClone,"droppable2":sourceClone};    
    }
};

const grid = 8;

export const getItemStyle = (isDragging:boolean, draggableStyle:any) => ({
    // some basic styles to make the items look a bit nicer
    userSelect: 'none',
    padding: grid * 2,
    margin: `0 0 ${grid}px 0`,

    // change background colour if dragging
    background: isDragging ? '#ffd166' : '#06d6a0',
    display:'flex',
    // styles we need to apply on draggables
    ...draggableStyle
});

export const getDoneItemStyle = (isDragging:boolean, draggableStyle:any) => ({
    // some basic styles to make the items look a bit nicer
    textDecoration: 'line-through',
    userSelect: 'none',
    padding: grid * 2,
    margin: `0 0 ${grid}px 0`,

    // change background colour if dragging
    background: isDragging ? '#ffd166' : '#407080',
    color: '#ffffff',
    display:'flex',
    // styles we need to apply on draggables
    ...draggableStyle
});

export const getListStyle = (isDraggingOver:boolean)=> ({
    background: isDraggingOver ? '#06d6a0' : '#efefef',
    padding: '2rem',
    width: '100%'
});
export const getDoneListStyle = (isDraggingOver:boolean)=> ({
    background: isDraggingOver ? 'lightblue' : '#efefef',
    padding: '2rem',
    width: '100%'
});


