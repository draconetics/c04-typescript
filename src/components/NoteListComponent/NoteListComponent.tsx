import React, { useEffect, useState} from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faExpandArrowsAlt, faPenSquare, faWindowClose } from '@fortawesome/free-solid-svg-icons'
import { DragDropContext, Droppable, Draggable, DropResult, DraggableProvided } from 'react-beautiful-dnd';
import AddNoteComponent from '../AddNoteComponent'
import {reorder, move, getListStyle, getDoneListStyle, getItemStyle, getDoneItemStyle} from './noteListUtility'

interface IProps {
    noteList: INote[],
    getNotes: () => void,
    error: string,
    loading: boolean,
    setNotes: (data:INote[])=> void
  }


export const NoteListComponent:React.FC<IProps> = (props)=>{
    const getDefaultNote = ()=> {
        return {
            _id:""+new Date().getTime(),
            text:"",
            complete:false
        }
    };
    const [editedNote, setEditedNote] = useState(getDefaultNote())
    const [editMode, setEditMode] = useState<boolean>(false);
    const [noteDoneList, setNoteDoneList] = useState<INote[]>([]);
    
    //warning with hooks
    const getNotes =  props.getNotes;
    useEffect(()=>{
        getNotes();
    },[getNotes])

    const editNote = (id:string)=>{
        const found = props.noteList.find(item => item._id === id);
        if(found){
            setEditedNote(found);
            setEditMode(true)
        }
            
    }

    const saveNote = ()=>{
        if(editedNote.text.trim() === "")
            return;
        if(editMode){
            const newList = props.noteList.map((item)=>{
                        if(item._id === editedNote._id)
                            item.text = editedNote.text
                        return item;
                    });
            props.setNotes(newList)
            setEditMode(false)
        }else{
            props.setNotes([...props.noteList, editedNote])
        }
        setEditedNote(getDefaultNote())
    }

    const deleteNote = (id:string)=>{
    
        let newList = props.noteList.filter(item=>item._id !== id?item:null);
        props.setNotes(newList)
        newList = noteDoneList.filter(item=>item._id !== id?item:null);
        setNoteDoneList(newList)
    }

    const cancel = ()=>{
        setEditedNote(getDefaultNote());
        setEditMode(false);
    }

    const showAlert = () =>{
        if(props.error){
            return <span className="alert alert-danger">{props.error}</span>
        }
        return null;
    }

    const showAlertEmptyList = (list:INote[]) =>{
        if(list.length === 0)
            return (<span>List is Empty</span>)
        else
            return null;
    }

    
    const droppableIdToList = (id:string) => {
        if(id === "droppable")
            return props.noteList;
        else
            return noteDoneList;
    }

    const onDragEnd = (result:DropResult) => {
        const { source, destination } = result;

        // dropped outside the list
        if (!destination) {
            return;
        }

        if (source.droppableId === destination.droppableId) {
            const orderedItems = reorder(
                droppableIdToList(source.droppableId),
                source.index,
                destination.index
            );
/*
            let state = { ...list, items:orderedItems };

            if (source.droppableId === 'droppable2') {
                state = { ...list, doneItems: orderedItems };
            }
            setList(state);
        */
            if (source.droppableId === 'droppable2') {
                setNoteDoneList(orderedItems)
            }else{
                props.setNotes(orderedItems)
            }
            
        } else {
            const result = move(
                droppableIdToList(source.droppableId),
                droppableIdToList(destination.droppableId),
                source,
                destination
            );
/*
            setList({
                items: result.droppable,
                doneItems: result.droppable2
            });*/
            props.setNotes(result.droppable)
            setNoteDoneList(result.droppable2)
        }
    };

    const noteBoardList = ()=>{
        return (<div className="note-board__list">
        <h2>TODO LIST</h2>
        <Droppable droppableId="droppable" data-test="droppable">
            {(provided, snapshot) => (<>
                
                <div
                    ref={provided.innerRef}
                    style={getListStyle(snapshot.isDraggingOver)}>
                    {showAlertEmptyList(props.noteList)} 
                    {props.noteList.map((item, index) => (
                        <Draggable
                            key={item._id}
                            draggableId={item._id}
                            index={index}>
                            {(provided, snapshot) => (
                                
                                <div
                                    ref={provided.innerRef}
                                    {...provided.draggableProps}
                                    {...provided.dragHandleProps}
                                    style={getItemStyle(
                                        snapshot.isDragging,
                                        provided.draggableProps.style
                                    )}>
                                    <FontAwesomeIcon icon={faExpandArrowsAlt} size="2x"/>
                                    <span className="note-board__item__text">{item.text}</span>
                                    <FontAwesomeIcon icon={faPenSquare}  size="2x" onClick={()=>editNote(item._id)} className="note-board__icon"/>
                                    <FontAwesomeIcon icon={faWindowClose}  size="2x" onClick={()=>deleteNote(item._id)} className="note-board__icon"/>
                                </div>
                                
                            )}
                        </Draggable>
                    ))}
                    {provided.placeholder}
                </div>
            </>)}
        </Droppable>
        </div>);
    }

    const noteBoardListDone = ()=>{
        return(<div className="note-board__list--done">
            <h2>DONE LIST</h2>
            <Droppable droppableId="droppable2" data-test="droppable2">
                {(provided, snapshot) => (
                    <div
                        data-test="div"
                        ref={provided.innerRef}
                        style={getDoneListStyle(snapshot.isDraggingOver)}>
                        {showAlertEmptyList(noteDoneList)}      
                        {noteDoneList.map((item, index) => (
                            <Draggable
                                key={item._id}
                                draggableId={item._id}
                                index={index}>
                                {(provided:DraggableProvided, snapshot) => (
                                    <div
                                        ref={provided.innerRef}
                                        {...provided.draggableProps}
                                        {...provided.dragHandleProps}
                                        style={getDoneItemStyle(
                                            snapshot.isDragging,
                                            provided.draggableProps.style
                                        )}>
                                        <span className="note-board__item__text">{item.text}</span>
                                        <FontAwesomeIcon icon={faWindowClose}  size="2x" onClick={()=>deleteNote(item._id)} className="note-board__icon"/>
                                    </div>
                                )}
                            </Draggable>
                        ))}
                        {provided.placeholder}
                    </div>
                )}
            </Droppable>
        </div>);
    }

    return (<div data-test="NoteListComponent">
            <div className="note-board container">

            <AddNoteComponent 
                editedNote={editedNote} 
                setEditedNote={setEditedNote} 
                editMode={editMode} 
                saveNote={saveNote}
                cancel={cancel}/>
            {showAlert()}
            <DragDropContext onDragEnd={onDragEnd}>
                {noteBoardList()}                
                {noteBoardListDone()}
            </DragDropContext>
            </div>


    </div>);
}
