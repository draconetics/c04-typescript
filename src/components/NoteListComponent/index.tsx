import {NoteListComponent} from './NoteListComponent'
import { connect } from 'react-redux';

import {getNotes} from '../../actions/noteAction'
import {SET_NOTES} from '../../actions/types'
//import * as actionCreators from '../../actions/noteAction'

/*
interface StateFromProps {
  noteList: INote[]
}

interface DispatchFromProps {
  getNotes: () => void;
}*/

const mapStateToProps = (state:any) =>{
      return {
          noteList: state.noteReducer.notes,
          error: state.noteReducer.error,
          loading: state.noteReducer.loading
      }
  }
 
const mapDispatchToProps = (dispatch: AppDispatch)=>{
    
    return {
      getNotes: () => dispatch(getNotes()),
      setNotes: (newNoteList:INote[]) => dispatch({type:SET_NOTES,value:newNoteList})
    };
}
  
export default connect(mapStateToProps, mapDispatchToProps)(NoteListComponent);