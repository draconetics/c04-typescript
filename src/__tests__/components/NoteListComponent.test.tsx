import { shallow } from 'enzyme';
import React from 'react';
import { NoteListComponent } from '../../components/NoteListComponent/NoteListComponent';

describe("#ListItemComponent",()=>{
    describe("checking PropTypes", ()=>{
        it('should NOT show a list because noteList:[]', ()=>{
            const properties = {
                noteList:[],
                getNotes: jest.fn(),
                error: ""
            }
            const appWrapper = shallow(<NoteListComponent {...properties}/>)
            const noteListComponent = appWrapper.find(`[data-test='NoteListComponent']`);
            const alertError = appWrapper.find(`.alert`);
            
            //render
            expect(noteListComponent).toHaveLength(1);
            //no errors
            expect(alertError).toHaveLength(0);
            //show message is empty
            expect(noteListComponent.text().includes('List is empty')).toBe(true);
            //do not show items
            expect(noteListComponent.contains('li')).toEqual(false);
        })

        it('should show a list because noteList:[1item]', ()=>{
            const properties = {
                noteList:[{_id:"123343455667abcdef",text:"this is a task", complete:false}],
                getNotes: jest.fn(),
                error: ""
            }
            const appWrapper = shallow(<NoteListComponent {...properties}/>)
            const noteListComponent = appWrapper.find(`[data-test='NoteListComponent']`);
            const alertError = appWrapper.find(`.alert`);
            
            //render
            expect(noteListComponent).toHaveLength(1);
            //no errors
            expect(alertError).toHaveLength(0);
            //do not show message is empty
            expect(noteListComponent.text().includes('List is empty')).toBe(false);
            //snapshot
            expect(noteListComponent).toMatchSnapshot();
            //do not show items
            expect(noteListComponent.text().includes('this is a task')).toBe(true);
            
        })

        it('should show an error; error:"Network error"', ()=>{
            const properties = {
                noteList:[],
                getNotes: jest.fn(),
                error: "Network error"
            }
            const appWrapper = shallow(<NoteListComponent {...properties}/>)
            const noteListComponent = appWrapper.find(`[data-test='NoteListComponent']`);
            const alertError = appWrapper.find('.alert');
            
            //render
            expect(noteListComponent).toHaveLength(1);
            //no errors
            expect(alertError).toHaveLength(1);
            //show message is empty
            expect(noteListComponent.text().includes('List is empty')).toBe(true);
            //do not show items
            expect(noteListComponent.contains('li')).toEqual(false);
        })

    })
})