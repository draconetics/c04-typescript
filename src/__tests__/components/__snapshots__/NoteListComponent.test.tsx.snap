// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#ListItemComponent checking PropTypes should show a list because noteList:[1item] 1`] = `
<div
  data-test="NoteListComponent"
>
  <div
    className="note-container"
  >
    <ul
      className="note-list"
      id="note-list"
      onDragOver={[Function]}
      onDrop={[Function]}
    >
      <li
        draggable="true"
        id="note-list-0"
        key="0"
        onDragStart={[Function]}
      >
        this is a task
      </li>
    </ul>
    <ul
      className="note-done-list"
      id="note-done-list"
      onDragOver={[Function]}
      onDrop={[Function]}
    />
  </div>
</div>
`;
